staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,239 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+// SizeLimitedDiskCache is an implementation of httpcache.Cache that supplements
+// the in-memory map with persistent storage up to a pre-defined maxSize. After
+// the disk cache size breaches the size limit, the oldest files (by file
+// modification times) are evicted until the size limit is satisfied. This cache
+// is meant to be a substitute for "github.com/gregjones/httpcache/diskcache".
+type SizeLimitedDiskCache struct {
+	maxSize     uint64
+	currentSize uint64
+	fileInfos   []fileInfo
+	diskv       *diskv.Diskv
+}
+
+// Get returns the response corresponding to key if present.
+func (c *SizeLimitedDiskCache) Get(key string) (resp []byte, ok bool) {
+	key = keyToFilename(key)
+	resp, err := c.diskv.Read(key)
+	if err != nil {
+		return []byte{}, false
+	}
+	return resp, true
+}
+
+// Set saves a response to the cache as key, and updates the stored
+// fileInfos and currentSize. This function can cause cache evictions.
+func (c *SizeLimitedDiskCache) Set(key string, resp []byte) {
+	key = keyToFilename(key)
+	size := uint64(len(resp))
+	c.trimDiskCache(size)
+	err := c.diskv.WriteStream(key, bytes.NewReader(resp), true)
+	if err == nil {
+		c.appendFileInfo(fileInfo{name: key, size: size, modTime: time.Now()})
+	} else {
+		klog.Errorf("Error disk cache entry (%s):  [%v]", key, err)
+	}
+}
+
+// Delete removes the response with key from the cache, updating the
+// caches fileInfos and the caches currentSize.
+func (c *SizeLimitedDiskCache) Delete(key string) {
+	key = keyToFilename(key)
+	err := c.diskv.Erase(key)
+	if err == nil {
+		c.removeFileInfo(key)
+	} else {
+		klog.Errorf("Error deleting entry (%s) from disk cache:  [%v]", key, err)
+	}
+}
+
+// Passing the size of a file that is going to be added, evicts entries
+// from the cache until the current cache size plus the passed size is below
+// the maximum cache disk size. Can update the cache fileInfos and currentSize
+// if evictions occur.
+func (c *SizeLimitedDiskCache) trimDiskCache(size uint64) {
+	for ((c.currentSize + size) > c.maxSize) && (len(c.fileInfos) > 0) {
+		fi := c.fileInfos[0] // Evict oldest file
+		err := c.diskv.Erase(fi.name)
+		if err == nil {
+			c.currentSize -= fi.size
+			c.fileInfos = c.fileInfos[1:]
+		} else {
+			// If disk cache eviction fails, just exit.
+			klog.Errorf("Error trimming disk cache: [%v]\n", err)
+			break
+		}
+	}
+}
+
+// IMPORTANT: This function MUST be the same as
+// "github.com/gregjones/httpcache/diskcache.keyToFilename()" in
+// order to maintain compatibility with existing http disk caches.
+func keyToFilename(key string) string {
```

Comment:
Have you tried writing a layer above or below `httpcache/diskcache` so that you wouldn't have to re-implement this function and/or the `Get` method (which is a strict copy-paste)? I'm curious if it would have avoided some duplication and/or implicit requirements.